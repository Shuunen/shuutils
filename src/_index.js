'use strict'

import { getTimestamp, getTimestampMs } from './timestamps'

module.exports = { getTimestamp, getTimestampMs }