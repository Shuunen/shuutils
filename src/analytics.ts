/* eslint-disable @typescript-eslint/prefer-readonly-parameter-types */
import { sleep } from './functions'
import { Logger } from './logger'

type AnalyticsInstance = {
  identify: (userId: string, additionalData?: Record<string, unknown>) => Promise<void>
  page: () => Promise<void>
  track: (event: string, additionalData?: Record<string, unknown>) => Promise<void>
}

/**
 * Generic analytics wrapper, can be used with any analytics third party library
 * @example const analytics = new Analytics()
 * @example analytics.setupInMemory('my-app')
 * @example analytics.track('invoice-download')
 * @example analytics.track('click', { category: '...', label: '...', ... })
 */
// eslint-disable-next-line no-restricted-syntax
export class Analytics {
  private $app = ''
  private $instance: AnalyticsInstance | undefined = undefined
  private readonly $logger = new Logger()
  public readonly pile: string[] = []

  /**
   * Identify a user
   * @param userId - The user, like '123456', 'super-admin', etc.
   * @param additionalData - Additional data to identify the user, like `{ email: '...', age: 35, ... }`
   */
  public identify(userId: string, additionalData?: Record<string, unknown>) {
    if (this.$instance === undefined) throw new Error('analytics : identify failed, analytics not setup')
    this.$logger.info(`analytics : identify user "${userId}" on app "${this.$app}"`)
    void this.$instance.identify(userId, additionalData)
  }

  /**
   * Track a page view
   */
  public page() {
    if (this.$instance === undefined) throw new Error('analytics : track page failed, analytics not setup')
    this.$logger.debug(`analytics : track page for app "${this.$app}"`)
    void this.$instance.page()
  }

  /**
   * Setup analytics with in memory tracking, useful for tests
   * @param app - The name of the app, like 'mlm-react-front', 'cosmos-front', etc.
   */
  public setupInMemory(app: string) {
    this.$app = app
    // eslint-disable-next-line jsdoc/require-jsdoc
    const addToPile = async (item: string, data?: Record<string, unknown>) => {
      this.$logger.debug(`analytics : add "${item}" to pile`, data)
      await sleep(1)
      this.pile.push(item)
      this.$logger.debug(`analytics : pile for app "${app}"`, this.pile)
    }
    this.$instance = {
      // eslint-disable-next-line jsdoc/require-jsdoc
      identify: async (userId: string, additionalData?: Record<string, unknown>) => {
        await addToPile(`identify:${userId}`, additionalData)
      },
      // eslint-disable-next-line jsdoc/require-jsdoc
      page: async () => {
        await addToPile('page')
      },
      // eslint-disable-next-line jsdoc/require-jsdoc
      track: async (event: string, additionalData?: Record<string, unknown>) => {
        await addToPile(`track:${event}`, additionalData)
      },
    }
    this.$logger.info(`analytics : setup app "${app}" with in memory tracking`)
  }

  /**
   * Track an event
   * @param event - The event name like 'click', 'download-invoice', etc.
   * @param additionalData - Additional data to track the event, like `{ category: '...', label: '...', ... }`
   */
  public track(event: string, additionalData?: Record<string, unknown>) {
    if (this.$instance === undefined) throw new Error('analytics : track event failed, analytics not setup')
    this.$logger.debug(`analytics : track event "${event}" for app "${this.$app}"`)
    void this.$instance.track(event, additionalData)
  }
}
