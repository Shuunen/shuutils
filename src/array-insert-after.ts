import { nbAfter, nbNoIndex } from './constants'
import { clone } from './object-clone'

/**
 * Add a value to an array if it doesn't already exist
 * @param array the array to add the value to
 * @param item the value to add
 * @param value the value to add
 * @returns the array with the value added (if it didn't already exist)
 */
export function insertValueAfterItem<T> (array: T[], item: T, value: T) {
  const index = array.indexOf(item)
  if (index === nbNoIndex) return array
  const arrayCopy = clone(array)
  const start = Math.min(index + nbAfter, arrayCopy.length)
  arrayCopy.splice(start, 0, value)
  return arrayCopy
}
