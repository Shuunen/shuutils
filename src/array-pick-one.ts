/**
 * Return an item from an array
 * @param items like : ["great", "place", "pine"]
 * @returns item like : "pine"
 */
export function pickOne<Type>(items: readonly Type[]) {
  if (items.length === 0) throw new Error('Array is empty')
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  return items[Math.floor(Math.random() * items.length)] as Type
}
