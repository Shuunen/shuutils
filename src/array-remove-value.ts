import { clone } from './object-clone'

/**
 * Remove a value from an array
 * @param array the array to remove the value from
 * @param value the value to remove
 * @returns a copy of the array without the value
 */
export function removeValueFromArray<Type>(array: readonly Type[], value: Type) {
  if (!array.includes(value)) return array
  const arrayCopy = clone(array)
  arrayCopy.splice(arrayCopy.indexOf(value), 1)
  return arrayCopy
}
