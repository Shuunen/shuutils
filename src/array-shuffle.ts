import { Nb } from './constants'
import { clone } from './object-clone'

/**
 * Randomize an array using the Fisher-Yates algorithm
 * source : https://dev.to/codebubb/how-to-shuffle-an-array-in-javascript-2ikj
 * @param input the array to shuffle
 * @returns a shuffled clone of the input array
 */
export function shuffleArray<T> (input: T[]) {
  const array = clone(input)
  // eslint-disable-next-line for-direction
  for (let indexA = array.length - 1; indexA > 0; indexA += Number(Nb.Previous)) {
    const indexB = Math.floor(Math.random() * (indexA + 1))
    const temporary = array[indexA]
    array[indexA] = array[indexB] as T // eslint-disable-line @typescript-eslint/consistent-type-assertions
    array[indexB] = temporary as T // eslint-disable-line @typescript-eslint/consistent-type-assertions
  }
  return array
}
