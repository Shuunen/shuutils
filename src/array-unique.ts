/**
 * Return an array with unique values
 * @param items the array to check/clean/reduce
 * @returns the array with unique values
 */
export function arrayUnique<Type>(items: readonly Type[]) {
  return Array.from(new Set(items))
}
