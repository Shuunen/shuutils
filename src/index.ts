export * from './array-insert-after'
export * from './array-pick-one'
export * from './array-remove-value'
export * from './array-shuffle'
export * from './array-unique'
export * from './browser-clipboard'
export * from './browser-fill-like-human'
export * from './browser-inject-styles'
export * from './browser-page-change'
export * from './browser-scout'
export * from './browser-toast'
export * from './colors'
export * from './constants'
export * from './date-relative'
export * from './dates'
export * from './debounce'
export * from './dom'
export * from './environment'
export * from './events'
export * from './expect-type'
export * from './functions'
export * from './logger'
export * from './memoize'
export * from './number-random'
export * from './number-ranged-score'
export * from './number-round'
export * from './object-cap-nesting'
export * from './object-clone'
export * from './object-equal'
export * from './object-flatten'
export * from './object-serializer'
export * from './object-sort'
export * from './objects'
export * from './state'
export * from './storage'
export * from './string-remove-accents'
export * from './strings'
export * from './testing-align-for-snap'
export * from './throttle'
export * from './timestamps'
export * from './types'
