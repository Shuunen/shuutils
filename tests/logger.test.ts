import { Logger, LogLevel, red } from '../src'
import { check } from './utils'

const loggerA = new Logger()
check('loggerA is active', loggerA.options.isActive, true)
loggerA.info('This info 0 should be logged')

const loggerB = new Logger({ isActive: false, willOutputToMemory: true, willOutputToConsole: false })
check('loggerB is not active', loggerB.options.isActive, false)
loggerB.info('This info 1 should not be logged')
loggerB.options.isActive = true
loggerB.info('This info 2 should be logged')
loggerB.info('This info 3 should be logged too')
loggerB.warn('This warn 1 should be logged')
check('loggerB 3 inMemoryLogs', loggerB.inMemoryLogs.length, 3)

const loggerC = new Logger({ willLogDate: true, willLogTime: true, willOutputToConsole: false, minimumLevel: LogLevel.Error, willOutputToMemory: true })
loggerC.warn('This warn 2 should not be logged')
loggerC.error('This error 1 should be logged')
loggerC.disable()
loggerC.error('This error 2 should not be logged')
loggerC.enable()
loggerC.error('This error 3 should be logged')
loggerC.test(true, 'This test 1 should not be logged')
loggerC.options.minimumLevel = LogLevel.Test
loggerC.test(true, 'This test 2 should be logged')
loggerC.test(false, 'This test 3 should be logged')
check('loggerC 4 inMemoryLogs', loggerC.inMemoryLogs.length, 4)

const loggerD = new Logger({ willLogDelay: false, willOutputToMemory: true, willOutputToConsole: false })
loggerD.info('This info 4 should be logged', 12)
loggerD.info('This info 5 should be logged too', [1, 2, 3])
loggerD.warn('This warn 3 should be logged', { keyA: 1, keyB: 'John', isKeyC: true })
loggerD.disable()
loggerD.warn('This warn 4 should not be logged')
loggerD.error('This error 4 should not be logged')
loggerD.enable()
loggerD.error('This error 5 should be logged', null) // eslint-disable-line unicorn/no-null
loggerD.test(true, 'This test 4 should be logged', undefined) // eslint-disable-line unicorn/no-useless-undefined
loggerD.test(false, 'This test 5 should be logged', () => 'Hello world')
loggerD.debug('This debug 1 should be logged', true, [], {})
loggerD.options.minimumLevel = LogLevel.Info
loggerD.debug('This debug 2 should not be logged')
check('loggerD 7 inMemoryLogs', loggerD.inMemoryLogs.length, 7)
check('loggerD inMemoryLogs', loggerD.inMemoryLogs, [
  ' info This info 4 should be logged 12',
  ' info This info 5 should be logged too [1,2,3]',
  ' warn This warn 3 should be logged {\'keyA\':1,\'keyB\':\'John\',\'isKeyC\':true}',
  'error This error 5 should be logged null',
  '   ✓  This test 4 should be logged undefined',
  '   ✗  This test 5 should be logged () => \'Hello world\'',
  'debug This debug 1 should be logged true [] {}',
])

const loggerE = new Logger({ willOutputToConsole: false })
check('loggerE clean A', loggerE.clean(), '')
check('loggerE clean B', loggerE.clean(red('Oh I\'m in red now ?!')), 'Oh I\'m in red now ?!')
check('loggerE clean C', loggerE.clean('an array ?', [12, 42]), 'an array ? [12,42]')
check('loggerE clean D', loggerE.clean('a function ?', () => 'Hello world'), 'a function ? () => \'Hello world\'')
check('loggerE clean E', loggerE.clean('an object ?', { keyA: 1, keyB: 'John', isFull: true }), 'an object ? {\'keyA\':1,\'keyB\':\'John\',\'isFull\':true}')
check('loggerE clean F', loggerE.clean('a boolean ?', true), 'a boolean ? true')
check('loggerE clean G', loggerE.clean('a number ?', 42), 'a number ? 42')
check('loggerE clean H', loggerE.clean('a string ?', 'Hello world'), 'a string ? Hello world')
check('loggerE clean I', loggerE.clean('a null ?', null), 'a null ? null') // eslint-disable-line unicorn/no-null
check('loggerE clean J', loggerE.clean('an undefined ?', undefined), 'an undefined ? undefined') // eslint-disable-line unicorn/no-useless-undefined
check('loggerE clean K', loggerE.clean('a date ?', new Date('2020-01-01')), 'a date ? \'2020-01-01T00:00:00.000Z\'')
check('loggerE clean L', loggerE.clean('a regexp ?', /Hello world{3,5}/u), 'a regexp ? {}') // not supported for now

