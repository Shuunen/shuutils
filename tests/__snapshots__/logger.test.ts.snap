// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`logger B 1`] = `
[
  " info This info 2 should be logged",
  " info This info 3 should be logged too",
  " good This success 0 should be logged",
  " warn This warn 1 should be logged",
]
`;

exports[`logger D 1`] = `
[
  " info This info 4 should be logged 12",
  " info This info 5 should be logged too [1,2,3]",
  " warn This warn 3 should be logged {'isKeyC':true,'keyA':1,'keyB':'John'}",
  "error This error 5 should be logged null",
  "   ✓  This test 4 should be logged undefined",
  "   ✗  This test 5 should be logged () => 'Hello world'",
  "debug This debug 1 should be logged true [] {}",
]
`;
